File: 그래프_dfs/프로그래머스lv3-여행경로.py
# 리트코드332 - 일정 재구성 문제와 동일
from collections import defaultdict, deque
def solution(tickets):
    graph = defaultdict(deque)
    for v, w in sorted(tickets):
        graph[v].append(w)
    
    routes = []
    
    def dfs(v):
        while graph[v]:
            dfs(graph[v].popleft())
        routes.append(v)
        
    dfs("ICN")
    
    return routes[::-1]
    
"""
정확성  테스트
테스트 1 〉	통과 (0.02ms, 10.2MB)
테스트 2 〉	통과 (0.01ms, 10.2MB)
테스트 3 〉	통과 (0.01ms, 10.2MB)
테스트 4 〉	통과 (0.01ms, 10.2MB)
"""

# 반복문과 스택으로 풀이
from collections import defaultdict
def solution(tickets):
    graph = defaultdict(list)
    for v, w in tickets:
        graph[v].append(w)
    
    for v in graph.keys():
        graph[v].sort(reverse = True)
    
    stack = ["ICN"]
    result = []
    
    while stack:
        v = stack[-1]
        if graph[v]:
            stack.append(graph[v].pop())
        else:
            result.append(stack.pop())
    
    return result[::-1]
"""
정확성  테스트
테스트 1 〉	통과 (0.02ms, 10MB)
테스트 2 〉	통과 (0.01ms, 10.2MB)
테스트 3 〉	통과 (0.01ms, 10MB)
테스트 4 〉	통과 (0.01ms, 10.2MB)
"""

File: 그래프_dfs/프로그래머스lv3-여행경로.js
// 데크가 아니라 큐만 사용한다면 거꾸로 스택으로 푸는 것도 가능.
function solution(tickets) {
  const graph = tickets
    .sort()
    .reverse()
    .reduce((obj, [v, w]) => {
      if (!obj[v]) obj[v] = []
      obj[v].push(w)
      return obj
    }, {})

  const result = []

  const dfs = (node) => {
    while (graph[node] && graph[node].length > 0) {
      dfs(graph[node].pop())
    }
    result.push(node)
  }

  dfs("ICN")

  return result.reverse()
}
// 정확성  테스트
// 테스트 1 〉	통과 (0.26ms, 33.5MB)
// 테스트 2 〉	통과 (0.14ms, 33.7MB)
// 테스트 3 〉	통과 (0.10ms, 33.6MB)
// 테스트 4 〉	통과 (0.11ms, 33.5MB)
